storage/innobase/include/dict0mem.h

```diff
@@ -1451,45 +1451,149 @@ typedef std::set<dict_v_col_t*, std::less<dict_v_col_t*>,
-					does not generate new indexes
-					implicitly */
-	dict_index_t*	referenced_index;/*!< referenced index */
-
-	dict_vcol_set*	v_cols;		/*!< set of virtual columns affected
-					by foreign key constraint. */
-
-	/** Check whether the fulltext index gets affected by
-	foreign key constraint */
-	bool affects_fulltext() const;
+struct dict_foreign_t
+{
+  /* Object is allocated from this memory heap */
+  mem_heap_t *heap;
+  /* id of the constraint as a null terminated string */
+  char       *id;
+  /* number of indexes first fields for which the foreign key
+  constraint is defined: We allow the indexes to contain more
+  fields than mentioned in the constraint, as long as the first
+  fields are as mentioned */
+  unsigned	n_fields:10;
+  /* 0 or DELETE_CASCADE OR DELETE_SET_NULL */
+  unsigned	type:6;
+  /* foreign table name */
+  char *foreign_table_name;
+  /* Foreign table name used for dict lookup */
+  char *foreign_table_name_lookup;
+  /* table where the foreign key is */
+  dict_table_t *foreign_table;
+  /* names of the columns in the foreign key */
+  const char  **foreign_col_names;
+  /* referenced table name */
+  char *referenced_table_name;
+  /* referenced table name for dict lookup */
+  char *referenced_table_name_lookup;
+  /* Table where the referenced key is */
+  dict_table_t *referenced_table;
+  /* Names of the referenced columns in the referenced table */
+  const char **referenced_col_names;
+  /* foreign index; we require that both tables contain explicitly
+  defined indexes for the constraint: InnoDB does not generate
+  new indexes implicitly */
+  dict_index_t *foreign_index;
+  /* referenced index */
+  dict_index_t *referenced_index;
+  /* set of virtual columns affected by foreign key constraint */
+  dict_vcol_set *v_cols;
+  /** Check whether the fulltext index gets affected by
+  foreign key constraint */
+  bool affects_fulltext() const;
+  /** The flags for ON_UPDATE and ON_DELETE can be ORed;
+  the default is that a foreign key constraint is enforced,
+  therefore RESTRICT just means no flag */
+  static constexpr unsigned DELETE_CASCADE= 1U;
+  static constexpr unsigned DELETE_SET_NULL= 2U;
+  static constexpr unsigned UPDATE_CASCADE= 4U;
+  static constexpr unsigned UPDATE_SET_NULL= 8U;
+  static constexpr unsigned DELETE_NO_ACTION= 16U;
+  static constexpr unsigned UPDATE_NO_ACTION= 32U;
+private:
+  /** Check whether the name exists in given column names
+  @retval offset or UINT_MAX if name not found */
+  unsigned col_exists(const char *name, const char **names) const
+  {
+    for (unsigned i= 0; i < n_fields; i++)
+    {
+      if (!strcmp(names[i], name))
+        return i;
+    }
+    return UINT_MAX;
+  }
+
+public:
+  /** Check whether the name exists in the foreign key column names
+  @retval offset in case of success
+  @retval UINT_MAX in case of failure */
+  unsigned col_fk_exists(const char *name) const
+  {
+    return col_exists(name, foreign_col_names);
+  }
```

Comment:
Let’s add `noexcept` to any member function that can’t throw exceptions (such as `std::bad_alloc` from C++ heap memory allocations). No member functions here should do that.